core/src/main/java/com/datastrato/gravitino/listener/PartitionEventDispatcher.java

```diff
@@ -0,0 +1,169 @@
+
+import com.datastrato.gravitino.NameIdentifier;
+import com.datastrato.gravitino.catalog.PartitionDispatcher;
+import com.datastrato.gravitino.exceptions.NoSuchPartitionException;
+import com.datastrato.gravitino.exceptions.PartitionAlreadyExistsException;
+import com.datastrato.gravitino.listener.api.event.AddPartitionEvent;
+import com.datastrato.gravitino.listener.api.event.AddPartitionFailureEvent;
+import com.datastrato.gravitino.listener.api.event.DropPartitionEvent;
+import com.datastrato.gravitino.listener.api.event.DropPartitionFailureEvent;
+import com.datastrato.gravitino.listener.api.event.ExistPartitionEvent;
+import com.datastrato.gravitino.listener.api.event.ExistPartitionFailureEvent;
+import com.datastrato.gravitino.listener.api.event.GetPartitionEvent;
+import com.datastrato.gravitino.listener.api.event.GetPartitionFailureEvent;
+import com.datastrato.gravitino.listener.api.event.ListPartitionEvent;
+import com.datastrato.gravitino.listener.api.event.ListPartitionFailureEvent;
+import com.datastrato.gravitino.listener.api.event.ListPartitionNamesEvent;
+import com.datastrato.gravitino.listener.api.event.ListPartitionNamesFailureEvent;
+import com.datastrato.gravitino.listener.api.event.PurgePartitionEvent;
+import com.datastrato.gravitino.listener.api.event.PurgePartitionFailureEvent;
+import com.datastrato.gravitino.listener.api.info.PartitionInfo;
+import com.datastrato.gravitino.rel.partitions.Partition;
+import com.datastrato.gravitino.utils.PrincipalUtils;
+
+/**
+ * {@code PartitionEventDispatcher} is a decorator for {@link PartitionDispatcher} that not only
+ * delegates partition operations to the underlying partition dispatcher but also dispatches
+ * corresponding events to an {@link com.datastrato.gravitino.listener.EventBus} after each
+ * operation is completed. This allows for event-driven workflows or monitoring of partition
+ * operations.
+ */
+public class PartitionEventDispatcher implements PartitionDispatcher {
+  private final EventBus eventBus;
+  private final PartitionDispatcher dispatcher;
+
+  /**
+   * Constructs a PartitionEventDispatcher with a specified EventBus and PartitionDispatcher.
+   *
+   * @param eventBus The EventBus to which events will be dispatched.
+   * @param dispatcher The underlying {@link PartitionDispatcher} that will perform the actual
+   *     partition operations.
+   */
+  public PartitionEventDispatcher(EventBus eventBus, PartitionDispatcher dispatcher) {
+    this.eventBus = eventBus;
+    this.dispatcher = dispatcher;
+  }
+
+  @Override
+  public Partition addPartition(NameIdentifier ident, Partition partition)
+      throws NoSuchPartitionException, PartitionAlreadyExistsException {
+    try {
+      Partition newPartition = dispatcher.addPartition(ident, partition);
+      eventBus.dispatchEvent(
+          new AddPartitionEvent(
+              PrincipalUtils.getCurrentUserName(), ident, new PartitionInfo(newPartition)));
+      return partition;
```

Comment:
you should return `newPartition` not `partition`,  and that's the reason why catalogIT fails. because the name of `Partition` is null, by the way, is this expected? @mchades

Reply:
yeah, `partitionName` is nullable. For example, in Hive catalog, user can only specific the parititon value and the partition name will be automatically generic by the catalog